x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/ReindexDataStreamIndexTransportAction.java

```diff
@@ -0,0 +1,240 @@
+                    listener.onResponse(null);
+                } else {
+                    var errorMessage = String.format(Locale.ROOT, "Could not set read-only on source index [%s]", sourceIndexName);
+                    listener.onFailure(new ElasticsearchException(errorMessage));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                if (e instanceof ClusterBlockException || e.getCause() instanceof ClusterBlockException) {
+                    // It's fine if read-only is already set
+                    listener.onResponse(null);
+                } else {
+                    listener.onFailure(e);
+                }
+            }
+        });
+    }
+
+    private void deleteDestIfExists(boolean doDelete, String destIndexName, ActionListener<AcknowledgedResponse> listener) {
+        if (doDelete == false) {
+            listener.onResponse(null);
+            return;
+        }
+
+        logger.info("Attempting to delete index [{}]", destIndexName);
+        var deleteIndexRequest = new DeleteIndexRequest(destIndexName).indicesOptions(IGNORE_MISSING_OPTIONS)
+            .masterNodeTimeout(TimeValue.MAX_VALUE);
+        var errorMessage = String.format(Locale.ROOT, "Failed to acknowledge delete of index [%s]", destIndexName);
+        client.admin().indices().delete(deleteIndexRequest, failIfNotAcknowledged(listener, errorMessage));
+    }
+
+    private void createIndex(IndexMetadata sourceIndex, String destIndexName, ActionListener<CreateIndexResponse> listener) {
+        logger.info("Creating destination index [{}] for source index [{}]", destIndexName, sourceIndex.getIndex().getName());
+
+        // Create destination with subset of source index settings that can be added before reindex
+        var settings = getPreSettings(sourceIndex);
+
+        var sourceMapping = sourceIndex.mapping();
+        Map<String, Object> mapping = sourceMapping != null ? sourceMapping.rawSourceAsMap() : Map.of();
+        var createIndexRequest = new CreateIndexRequest(destIndexName).settings(settings).mapping(mapping);
+
+        client.admin().indices().create(createIndexRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(CreateIndexResponse response) {
+                if (response.isAcknowledged()) {
+                    listener.onResponse(null);
+                } else {
+                    var errorMessage = String.format(Locale.ROOT, "Could not create index [%s]", destIndexName);
+                    listener.onFailure(new ElasticsearchException(errorMessage));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                // If dest index already exists, just reindex into the existing one
+                if (e instanceof ResourceAlreadyExistsException || e.getCause() instanceof ResourceAlreadyExistsException) {
+                    listener.onResponse(null);
```

Comment:
This is probably worthy of a debug-level log message -- I could see that being helpful while troubleshooting.